src/storage/v3/shard_rsm.cpp

```diff
@@ -510,9 +514,125 @@ msgs::WriteResponses ShardRsm::ApplyWrite(msgs::CommitRequest &&req) {
-  return msgs::GetPropertiesResponse{};
+msgs::ReadResponses ShardRsm::HandleRead(msgs::GetPropertiesRequest &&req) {
+  if (req.vertices_and_edges.empty()) {
+    return msgs::GetPropertiesResponse{.result = msgs::GetPropertiesResponse::FAILURE};
+  }
+
+  auto shard_acc = shard_->Access(req.transaction_id);
+  auto dba = DbAccessor{&shard_acc};
+  const auto view = storage::v3::View::NEW;
+
+  auto collect_props = [](const VertexAccessor &acc, const std::vector<PropertyId> &props, View view,
+                          const std::optional<EdgeAccessor> &e_acc) mutable -> std::optional<GetPropElement> {
+    std::vector<TypedValue> properties;
+    std::vector<PropertyId> ids;
+    for (const auto &prop : props) {
+      ShardResult<PropertyValue> result{PropertyValue()};
+      if (e_acc) {
+        result = e_acc->GetProperty(prop, view);
+      } else {
+        result = acc.GetProperty(prop, view);
+      }
+      if (result.HasError() && result.GetError() == common::ErrorCode::NONEXISTENT_OBJECT) {
+        continue;
+      }
+      if (result.HasError()) {
+        spdlog::debug("Encountered an Error while trying to get a vertex property.");
+        return std::nullopt;
+      }
+      properties.push_back(PropertyToTypedValue<TypedValue>(result.GetValue()));
+      ids.push_back(prop);
+    }
+    GetPropElement element{std::move(properties), std::move(ids), acc, e_acc};
+    return {std::move(element)};
+  };
+
+  auto find_edge = [](const VertexAccessor &v, const EdgeTypeId &e) -> std::optional<EdgeAccessor> {
+    auto in = v.InEdges(view, {e});
+    MG_ASSERT(in.HasValue());
+    for (auto &edge : in.GetValue()) {
+      if (edge.EdgeType() == e) {
+        return edge;
+      }
+    }
+
+    auto out = v.OutEdges(view, {e});
+    MG_ASSERT(out.HasValue());
+    for (auto &edge : out.GetValue()) {
+      if (edge.EdgeType() == e) {
+        return edge;
+      }
+    }
+    return std::nullopt;
+  };
+
+  std::vector<GetPropElement> elements;
+
+  for (const auto &[vertex, maybe_edge] : req.vertices_and_edges) {
+    const auto &[label, pk_v] = vertex;
+    auto pk = ConvertPropertyVector(pk_v);
+    auto v_acc = dba.FindVertex(pk, view);
+    if (!v_acc) {
+      return msgs::GetPropertiesResponse{.result = msgs::GetPropertiesResponse::OUT_OF_SHARD_RANGE};
+    }
+    std::optional<EdgeAccessor> e_acc;
+    if (maybe_edge) {
+      e_acc = find_edge(*v_acc, *maybe_edge);
+      if (!e_acc) {
+        return msgs::GetPropertiesResponse{.result = msgs::GetPropertiesResponse::OUT_OF_SHARD_RANGE};
+      }
+    }
+
+    const auto *symbol = (maybe_edge) ? expr::identifier_edge_symbol : expr::identifier_node_symbol;
+    if (req.filter && !FilterOnVertex(dba, *v_acc, {*req.filter}, symbol, e_acc)) {
+      continue;
+    }
+
+    std::optional<GetPropElement> collected_properties;
+    collected_properties = collect_props(*v_acc, req.property_ids, view, e_acc);
```

Comment:
Why do we collect the properties first and order afterwards? I don't see its advantage, but a possible big memory overhead.